File: HLSDownloader/HLS/AttributeListParser.swift
import Foundation

enum AttributeListParser {
    static func parse(_ text: String) -> [String: String] {
        var result: [String: String] = [:]
        var fields: [String] = []
        var current = ""
        var insideQuotes = false

        for character in text {
            if character == "\"" {
                insideQuotes.toggle()
                current.append(character)
            } else if character == "," && !insideQuotes {
                fields.append(current)
                current = ""
            } else {
                current.append(character)
            }
        }
        if !current.isEmpty {
            fields.append(current)
        }

        for field in fields {
            guard let equals = field.firstIndex(of: "=") else { continue }
            let key = field[..<equals].trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
            var value = field[field.index(after: equals)...].trimmingCharacters(in: .whitespacesAndNewlines)
            if value.count >= 2, value.first == "\"", value.last == "\"" {
                value.removeFirst()
                value.removeLast()
            }
            result[key] = value
        }
        return result
    }
}

File: android/app/src/main/java/com/example/hlsdownloader/core/AttributeListParser.kt
package com.example.hlsdownloader.core

object AttributeListParser {
    fun parse(text: String): Map<String, String> {
        val fields = mutableListOf<String>()
        val current = StringBuilder()
        var insideQuotes = false
        text.forEach { character ->
            when {
                character == '"' -> {
                    insideQuotes = !insideQuotes
                    current.append(character)
                }
                character == ',' && !insideQuotes -> {
                    fields += current.toString()
                    current.clear()
                }
                else -> current.append(character)
            }
        }
        if (current.isNotEmpty()) fields += current.toString()

        return buildMap {
            fields.forEach { field ->
                val equals = field.indexOf('=')
                if (equals < 0) return@forEach
                val key = field.substring(0, equals).trim().uppercase()
                var value = field.substring(equals + 1).trim()
                if (value.length >= 2 && value.first() == '"' && value.last() == '"') {
                    value = value.substring(1, value.length - 1)
                }
                put(key, value)
            }
        }
    }
}
